pkg/placement/placement.go

```diff
@@ -352,51 +313,125 @@ func (p *Service) ReportDaprStatus(stream placementv1pb.Placement_ReportDaprStat
-				if isActorRuntime {
-					p.membershipCh <- hostMemberChange{
-						cmdType: raft.MemberRemove,
-						host:    raft.DaprHostMember{Name: registeredMemberID},
-					}
-				}
 			} else {
-				// no actions for hashing table. Instead, MembershipChangeWorker will check
-				// host updatedAt and if now - updatedAt > p.faultyHostDetectDuration, remove hosts.
 				log.Debugf("Stream connection is disconnected with the error: %v", err)
 			}
 
+			if isActorRuntime {
+				p.membershipCh <- hostMemberChange{
+					cmdType: raft.MemberRemove,
+					host:    raft.DaprHostMember{Name: registeredMemberID, Namespace: namespace},
+				}
+			}
+
 			return nil
 		}
 	}
 
 	return status.Error(codes.FailedPrecondition, "only leader can serve the request")
 }
 
-// addStreamConn adds stream connection between runtime and placement to the dissemination pool.
-func (p *Service) addStreamConn(conn placementGRPCStream) {
-	p.streamConnPoolLock.Lock()
-	p.streamConnPool = append(p.streamConnPool, conn)
-	p.streamConnPoolLock.Unlock()
+func (p *Service) upsertRequired(namespace string, req *placementv1pb.Host) bool {
+	state := p.raftNode.FSM().State()
+	state.Lock.RLock()
+	defer state.Lock.RUnlock()
+
+	members, err := state.Members(namespace)
+	if err != nil {
+		// Couldn't find the namespace, so we need to upsert
+		return true
+	}
+
+	if m, ok := members[req.GetName()]; ok {
+		// If all attributes match, no upsert is required
+		return !(m.AppID == req.GetId() && m.Name == req.GetName() && cmp.Equal(m.Entities, req.GetEntities()))
+	}
+
+	return true
+}
+
+func (p *Service) validateClient(stream placementv1pb.Placement_ReportDaprStatusServer) (*spiffe.Parsed, error) {
+	sec, err := p.sec.Handler(stream.Context())
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "")
+	}
+
+	if !sec.MTLSEnabled() {
+		return nil, nil
+	}
+
+	clientID, ok, err := spiffe.FromGRPCContext(stream.Context())
+	if err != nil || !ok {
+		log.Debugf("failed to get client ID from context: err=%v, ok=%t", err, ok)
+		return nil, status.Errorf(codes.Unauthenticated, "failed to get client ID from context")
+	}
+
+	return clientID, nil
 }
 
-func (p *Service) deleteStreamConn(conn placementGRPCStream) {
-	p.streamConnPoolLock.Lock()
-	for i, c := range p.streamConnPool {
-		if c == conn {
-			p.streamConnPool = append(p.streamConnPool[:i], p.streamConnPool[i+1:]...)
-			break
-		}
+func (p *Service) receiveAndValidateFirstMessage(stream placementv1pb.Placement_ReportDaprStatusServer, clientID *spiffe.Parsed) (*placementv1pb.Host, error) {
+	firstMessage, err := stream.Recv()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to receive the first message: %v", err)
 	}
-	p.streamConnPoolLock.Unlock()
+
+	if clientID != nil && firstMessage.GetId() != clientID.AppID() {
+		return nil, status.Errorf(codes.PermissionDenied, "client ID %s is not allowed", firstMessage.GetId())
```

Comment:
I think we should add both the client mTLS ID _and_ the requested ID in the error string here.